### PR TITLE
Extend TopMetrics to support BucketSelector

### PR DESCRIPTION
**Pull request by kkrik-es**
TopMetrics aggregation currently requires explicitly providing the name of the "leaf" metric to be returned. When combined with a bucket filtering aggregation like BucketSelector, the latter uses the TopMetric aggregation name instead of the leaf metric to retrieve the field to apply filtering to. 

To support this, `BucketHelpers.resolveBucketValue` is extended to retrieve the value of a InternalMultiValueAggregation is applicable. This can only be applied if the size in the TopMetrics spec is set to '1', otherwise it's not possible to specify which value to use for selection.

Related to #73429.

### CODE REVIEW

**Comment by kkrik-es** [comment]
[This](
https://gradle-enterprise.elastic.co/s/xezaidpqmjakg/tests/:qa:mixed-cluster:v8.8.0%23mixedClusterTest/org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT/test%20%7Bp0=aggregations%2Fpipeline%2FMax%20pipeline%20on%20terms%20agg%7D?top-execution=1)  looks like something to fix: 

```
Expected: \[LongTerms\] at aggregation \[the_terms_2\] |  
-- | --
  | but: was "{root_cause=[], type=search_phase_execution_exception, reason=, phase=fetch, grouped=true, failed_shards=[], caused_by={type=illegal_argument_exception, reason=Can't sort a [lterms] aggregation [the_terms_2], ..
```

This comes from an InternalAggregation that has no sortValue, hiding a more informative error msg. Looking.

**Comment by kkrik-es** [comment]
Tests pass now.

**Comment by kkrik-es** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/analytics/top_metrics.yml [hunk 1]
```diff
@@ -757,3 +757,108 @@
...
+                  top_metrics:
+                    metrics:
+                      field: salary
+                    sort:
+                      birth_date: asc
+                    size: 1
+                having.top_salary:
+                  bucket_selector:
+                    buckets_path:
+                      ts: top_salary[salary]
```
Comment (martijnvg) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/analytics/top_metrics.yml:857]: maybe also add a test where a field is selected that doesn't exist and where the top_salary bucket is empty?

**Review by martijnvg — COMMENTED** [review]
This looks good! I left a few comments.

File: server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java [hunk 2]
```diff
@@ -214,6 +216,14 @@ public static Double resolveBucketValue(
                     } catch (NumberFormatException ex) {
                         throw formatResolutionError(agg, aggPathAsList, propertyValue);
                     }
+                } else if (propertyValue instanceof InternalMultiValueAggregation) {
+                    try {
+                        value = ((InternalMultiValueAggregation) propertyValue).sortValue(aggPathAsList.get(aggPathAsList.size() - 1))
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketHelpers.java]: Well, after understanding the logic a bit better while debugging TopHits, I think this works without any changes - the YAML test passes if we properly set the buckets_path. I'll update the issue and revert this PR. The YAML change will be added to a separate PR for TopHits.

**Comment by martijnvg** [comment]
In my mind I mixed up top_metrics and top_hits aggregation. For the top_hits aggregation this support is definitely missing and that is what I think the original issue is about. I think we do want to support `top_hits` aggregation for pipeline aggs like bucket selector. A search request like this one fails now:

```
GET /index1/_search
{
    "size": 0,
    "aggs": {
        "groupby": {
            "terms": {
                "field": "gender"
            },
            "aggs": {
                "top_salary": {
                    "top_hits": {
                        "size": 1,
                        "sort": [
                            {
                                "birth_date": {
                                    "order": "asc"
                                }
                            }
                        ],
                        "_source": {
                            "includes": [
                                "salary"
                            ]
                        }
                    }
                },
                "having.top_salary": {
                    "bucket_selector": {
                        "buckets_path": {
                            "hits": "top_salary[hits]"
                        },
                        "script": "params.hits.hits.0._souce.salary > 6000"
                    }
                }
            }
        }
    }
}
```

**Comment by kkrik-es** [comment]
Correct, I'll be sending a PR for `top_hits` shortly - need to complete testing.

**Pull request closed without merging**